docs/src/main/paradox/user/lang/java/crdt.md

```diff
@@ -1,6 +1,141 @@
+
+There are multiple ways that a CRDT entity may access its CRDT. It may have the CRDT injected directly into its constructor or a command handler - the value can be wrapped in an `Optional` to distinguish between entities that have been created and CRDTs that have not yet been created. If not wrapped in `Optional`, the CRDT will be created automatically, according to its type. The CRDT can also be read from any @javadoc[`CrdtContext`](io.cloudstate.javasupport.crdt.CrdtContext) via the @javadoc[`state`](io.cloudstate.javasupport.crdt.CrdtContext#state-java.lang.Class-) method.
+
+An entities CRDT can be created from the entities constructor using the `CrdtFactory` methods on @javadoc[`CrdtCreationContext`](io.cloudstate.javasupport.crdt.CrdtCreationContext), or using the same methods in a command handler using the @javadoc[`CommandContext`](io.cloudstate.javasupport.crdt.CommandContext). Note that the CRDT may only be created once, and only if it hasn't been provided by the CloudState proxy already. Any attempt to create a CRDT when one already exists will throw an `IllegalStateException`.
+
+For most use cases, simply injecting the CRDT directly into the constructor, and storing in a local field, will be the most convenient and straight forward method of using a CRDT. In our shopping cart example, we're going to use an @javadoc[`LWWRegisterMap`](io.cloudstate.javasupport.crdt.LWWRegisterMap), this shows how it may be injected:
+
+@@snip [ShoppingCartEntity.java](/docs/src/test/java/docs/user/crdt/ShoppingCartEntity.java) { #creation }
+
+In addition to the CRDT, the constructor may accept a @javadoc[`CrdtCreationContext`](io.cloudstate.javasupport.crdt.CrdtCreationContext).
+
+## Handling commands
+
+Command handlers can be declared by annotating a method with @javadoc[`@CommandHandler`](io.cloudstate.javasupport.crdt.CommandHandler). They take a context class of type @javadoc[`CommandContext`](io.cloudstate.javasupport.crdt.CommandContext).
+
+By default, the name of the command that the method handles will be the name of the method with the first letter capitalized. So, a method called `getCart` will handle gRPC service call command named `GetCart`. This can be overridden by setting the `name` parameter on the `@CommandHandler` annotation.
+
+The command handler also can take the gRPC service call input type as a parameter, to receive the command message. This is optional, sometimes it's not needed, for example, our `GetCart` service call doesn't need any information from the message, since it's just returning the current state as is. Meanwhile, the `AddItem` service call does need information from the message, since it needs to know the product id, description and quantity to add to the cart.
```

Comment:
Perhaps we should document what happens if there are missing command implementations, or if there are multiple with the same "name" in the annodation?

Reply:
If two have the same name, the user function will fail to start, with an error saying `Multiple methods found for handling command of name foo: methodOneName, methodTwoName`. If you define a method for a command that doesn't exist, you get this message: `Command handler method someMethodName for command SomeCommandName found, but the service has no command by that name.` And if a method is missing, you get this error message: `No command handler found for command SomeCommandName on CRDT entity: some.ClassName`. I don't think this needs to be documented, the error messages are completely clear, there's nothing the user needs to do that they couldn't work out themselves from the error messages, the documentation already describes the correct way of configuring it, there's no need for them to anticipate the errors ahead of time (eg they can't write any code to recover from it, the user function just refuses to start), and a single test that starts the entity will identify all the errors (though we do need to write a testkit so they can do this without a full cloudstate install).